src/box/relay_stream.c

```diff
@@ -0,0 +1,50 @@
+#include "relay_stream.h"
+#include "tweaks.h"
+#include "coio.h"
+#include "errinj.h"
+
+size_t relay_stream_flush_size = 16384;
+TWEAK_UINT(relay_stream_flush_size);
+
+void
+relay_stream_write(struct relay_stream *stream, const struct xrow_header *row)
+{
+	size_t used = region_used(&fiber()->gc);
+	int iovcnt;
+	struct iovec iov[XROW_IOVMAX];
+	xrow_to_iovec(row, iov, &iovcnt);
+	size_t len = 0;
+	for (int i = 0; i < iovcnt; i++) {
+		len += iov[i].iov_len;
+	}
+	char *data = xlsregion_alloc(&stream->lsregion, len, ++stream->lsr_id);
+
+	for (int i = 0; i < iovcnt; i++) {
+		memcpy(data, iov[i].iov_base, iov[i].iov_len);
+		data += iov[i].iov_len;
+	}
+	region_truncate(&fiber()->gc, used);
+}
+
+int
+relay_stream_flush(struct relay_stream *stream, struct iostream *io)
+{
+	int64_t to_flush = stream->lsr_id;
+	/*
+	 * Might flush more than requested if data is added to the buffer
+	 * during the coio_writev yield.
+	 */
+	while (to_flush > stream->flush_pos.max_id) {
+		struct iovec iov[IOV_MAX];
+		int iovcnt = lengthof(iov);
+		lsregion_to_iovec(&stream->lsregion, iov, &iovcnt,
+				  &stream->flush_pos);
+		if (coio_writev(io, iov, iovcnt, 0) < 0)
+			return -1;
+		lsregion_gc(&stream->lsregion, stream->flush_pos.max_id);
+	}
```

Comment:
I don't remember - we do have a guarantee, that only one fiber is here at a time, right? Otherwise 1) we get mixed garbage in the socket, 2) `lsregion_gc()` must use the position remembered before send. For (2) consider this: one  fiber writes to the buffer, takes iovec, moves `flush_pos`, and yields in `coio_writev()`, then another fiber does same. Then first fiber finishes and calls `lsregion_gc()` on the position updated by the second fiber. Hence deleting the data which is being sent right now in the second fiber.

Reply:
Thanks for paying attention!

Yes, we do have such guarantee. There's a single fiber which reads WALs and sends them to the remote peer, sends occasional heartbeats and processes RAFT pushes. Besides, it always flushes a full transaction (see `relay_send_tx()`).
If this wasn't the case (either the tx was sent in parts or multiple fibers wrote to the socket), we would still need the check `relay->is_sending_tx`, which got deleted some time ago.

BTW, when implementing the relay rework in scope of #9968, you'll have to pay attention that a full transaction is added to the stream wthout yields for the same reasons as mentioned above.